### ISSUE TITLE
Add an HTML-formatted constructions support into Razor instead of the removed @helper directive.

### ISSUE THREAD

**Issue by alexaku** [aspnet/Razor#715]
The @helper directive was removed from Razor (MVC 6) and it was not given any simple replacement. See https://github.com/aspnet/Razor/issues/281.
For example, if I have define HTML code `<a src="http://www.youtube.com/watch?v=4EGDxkWoUOY" title="click me"><b>MVC 6</b> documentation</a>` and I would like to use it in many places of my Razor-compatible web-page, then I must create a new file - a "partial view". The entire file for one HTML-formatted string? Furthermore, this separated file is for a single web-page only. What if there are many chunks of HTML-code, that I need to reuse in a single web-page? How many "partial view" files should I spawn? How convenient to accompany this zoo? The answer is obvious.

MVC developers have provided us with TagHelpers that are going to be added to ASP.NET MVC 6. Each tag helper is a new separate file (class). Furthermore, each of these separate file can't contain an HTML markup (Razor). So, it is not a replacement for @helpers. If you remember ViewComponents, then you understand that each new ViewComponent will add two new files - "class" and "view". It is not a replacement for @helpers too.

How can I reuse some portion of the simple HTML code in my "Razor file" without adding a bunch of extra files?

Dear MVC-developers, if you retain the @functions directive in the Razor in spite of deletion of the @helper directive, then you must add HTML markup feature in their bodies (@functions). See below two @functions that could produce the identical result (if you will develop the interpretator of inlined HTML blocks).

`@functions {`

`public HtmlString GetTableHeader1(String className)`
`{`
`HtmlString result = <text>`
`<tr class="@className">`
`<td title="abc">one</td>`
`<td title="xyz">two</td>`
`</tr>`
`</text>;`
`return result;`
`}`

`public HtmlString GetTableHeader2(String className)`
`{`
`HtmlString result = new HtmlString(@"`
`<tr class=""" + Html.Encode(className) + @""">`
`<td title=""abc"">one</td>`
`<td title=""xyz"">two</td>`
`</tr>`
`");`
`return result;`
`}`

`}`

**Comment by Eilon** [comment]
First, I should clarify that View Components were designed as a replacement for MVC 5.x's Child Actions, it is not at all meant to be a replacement for `@helper` functions.

It's also worth noting that `@helper` was never well-supported in MVC because the code that was generated for them was designed for ASP.NET Web Pages, not ASP.NET MVC. This meant that all the HTML helpers, context types, etc. that were available from an `@helper` function were not compatible with MVC. E.g. you couldn't use Model State or render an MVC HTML helper from one of them.

I think in the meantime there are several alternatives to `@helper` functions that are available in ASP.NET Core MVC, each with some pros and cons:

1. Use a view component
 * **Pros:** Supports strongly-typed parameters
 * **Cons:** If using a Razor CSHTML file, it means an extra CSHTML file for each set of HTML
2. Use an HTML partial view
 * **Pros:** Each view is a single file, can mix code + Razor/HTML markup in the same file
 * **Cons:** No strongly-typed parameters
3. Use an HTML helper or Razor tag helper
* **Pros:** Get rich Intellisense in Visual Studio; tag helpers can compose with each other and control allowed markup in the parser
 * **Cons:** Everything is code, so no Razor syntax within the helper

In summary, there are several rich alternatives, though there aren't any that are an exact replacement for `@helper` functions.

That's certainly a feature we will look at bringing back, but without the limitations that `@helper` functions had in MVC 5.x. At this time we simply don't have the time to design and implement this feature and bring it to the necessary quality level to ship it in this version.

**Comment by alexaku** [comment]
I have not asked about recovery of the @helper directive. I asked how reuse a chunk of HTML (Razor) code without adding a new file? It is impossible in MVC 6 RC, is't it?

**Comment by NTaylorMullen** [comment]
@alexaku you can do:

```
Func<dynamic, IHtmlContent> foo = @<p>Some HTML</p>;

@foo(null)
```

Which generates (behind the scenes):
```C#
Func<dynamic, IHtmlContent> foo = item => new HelperResult(async(__razor_template_writer) => {
  WriteLiteralTo(__razor_template_writer, "<p>Some HTML</p>");
});
```

If you want to get fancy and have it take a value by utilizing the generated `item` property and do:
```
Func<int, IHtmlContent> foo = @<p>The number you entered is: @item</p>;

@foo(1234)
```

**Comment by alexaku** [comment]
1) Thank you, @NTaylorMullen. IHtmlContent is a new part of the ASP.NET 5, but I don't found any documentation about this techniques of Razor markup. It is necessary to bring this information to the general public.
2) This still does not solve the problem of a parameterized Razor code. What if there are two or three parameters of different types?

**Comment by NTaylorMullen** [comment]
> Thank you, @NTaylorMullen. IHtmlContent is a new part of the ASP.NET 5, but I don't found any documentation about this techniques of Razor markup. It is necessary to bring this information to the general public.

Our documentation is a work in progress but rest assured it's being worked on. These are definitely a less known part of Razor.

> This still does not solve the problem of a parameterized Razor code. What if there are two or three parameters of different types?

As unfortunate as it might be you'd need to provide some sort of poco object in that case. A potential, less clean solution would be using variables from outside of the `@<p>...</p>`:

```
var firstName = "John";
var lastName = "Doe";

Func<int, IHtmlContent> person = @<p>@firstName @lastName is @item years old.</p>;

@person(30)
```

**Comment by alexaku** [comment]
Thank you for your patience, @NTaylorMullen. What do you think about the ability of implementation of this Razor syntax in the @functions directive?

`@functions`
`{`

`public IHtmlContent person(String firstName, String lastName, Int32 age)`
`{`
`Func<dynamic, IHtmlContent> result = @<p>@firstName @lastName is @age years old.</p>;`
`return result(null);`
`}`

`}`

`@person("John", "Doe", 30)`

**Comment by crbranch** [comment]
@Eilon Which of the alternatives you listed (if any) would be most appropriate for rendering a recursive data structure (e.g., treeview).  Example using the old @helper syntax here:

http://stackoverflow.com/a/6423891/333127

**Comment by Eilon** [comment]
I'll let @NTaylorMullen answer that, as he's the resident Razor expert.

**Comment by alexaku** [comment]
To @crbranch, @Eilon. 
@NTaylorMullen suggests using the following structure:
`@{`
`	Func<IEnumerable<Foo>, IHtmlContent> ShowTree = @<text>@{`
`		var foo = item;`
`		//insert your Razor code here...`
`	}</text>;`
`}`

`@ShowTree(new List<Foo>{...})`

**Comment by Eilon** [comment]
Ah, that looks fine to me, then.

**Comment by alexaku** [comment]
To @Eilon.
But it is less convenient than the @helper directive. Will we see the @helper directive in the MVC 6 RTM?

**Comment by Eilon** [comment]
@alexaku it will not be in ASP.NET Core 1.0 MVC at RTM (formerly known as MVC 6). But it's certainly a feature we will look to add back in the future.

**Comment by alexaku** [comment]
@Eilon,
What's the point to re-implement the @helper directive? You can incorporate `@<text>...</text>` in the @function directive.

**Comment by Eilon** [comment]
@alexaku sorry I might be confused. I thought you were asking for `@helper` to be added back, no?

One difference between `@helper` and stuff in `@functions` is that helpers are available in other files, but functions are local to the file they are in.

**Comment by alexaku** [comment]
I need the @helper directive for a seamless project migration from MVC 5 to MVC 6 (ASP.NET Core 1.0). MVC 6 has made backwards incompatible. There are too many the @helper directive usages in my projects. After migration, I won't need the old Razor functionality and will be ready to use any new Razor constructions. I think that the use of the @helper functions in other files must be replaced with the use of the View Component. For this case, the "View Component" is most suitable.

**Comment by alexaku** [comment]
@Eilon,
I want to say that, despite all the advantages, the design of the View Component invocation weird.
Syntactically correct call looks like this:
`@Component.Invoke(nameof(SomeComponentClass))`

**Comment by tboby** [comment]
@Eilon,
My common usecase for `@helper` is where I need markup templates that cover more than one property of the model. Say I want to divide my entire form into sets of four, with controls and validation displayed for each set. Viewmodels are impracticable as they result in vast duplication of metadata.

- Editor Templates: Don't accept multiple models
- Partial views: Lose model metadata/attributes
- Code helper: String/tag builders for the vast majority of common html structures, difficult to maintain

And from my understanding of view components, they'd suffer from the same duplication of model metadata as simply using viewmodels with editor templates.

`@helper`s seem to be the only way to make _html presentation_ templates without losing metadata or the benefits of using razor completely!

**Comment by Eilon** [comment]
Hmm, what model metadata gets lost? There's only one model metadata system and it should get picked up no matter whether it's a partial view, `@helper`, or anything else.

**Comment by alexaku** [comment]
@tboby,
I do not quite understand you too. Can you give an example of the code? May you give an example of the code?

**Comment by alexaku** [comment]
In the end, I propose to call it the "ASP.NET Core **0.6** MVC" until the old functionality  will be restored.

**Comment by AlekseyMartynov** [comment]
Same here.

We used `@helper` as a workaround for inability to nest `@<text>` tags.

Example:

**What we wanted:**
```
@Html.CoolStuff(@<text>

        @Html.CoolStuff(@<text>
            Nested
        </text>)

</text>);
```

**What we did in MVC 5:**
```
@helper NestedCoolStuff() {
    @Html.CoolStuff(@<text>
        Nested
    </text>)
}

@Html.CoolStuff(@<text>
   @NestedCoolStuff()
</text>);
```

**What we do in Core:**
:sob:

**Comment by Eilon** [comment]
@DamianEdwards @rynowak @NTaylorMullen have been looking at some possible improvements in this area, so there is certainly a positive outlook that we'll see improvements here, though we don't yet have any commitment or final design.

**Comment by binki** [comment]
@alexaku You’re making it look like it’s impossible to pass parameters to functions. I don’t know how to use MVC Core, but with the real version I can get typed parameters this way:

```razor
@{
Func<string, int, IEnumerable<string>, HelperResult> showThing = (name, age, tags) => new Func<object, HelperResult>(@<text>
<div>
  <h3>@name</h3>
  <p>Age: @age</p>
  <ul>
    @foreach (var tag in tags)
    {
      <li>@tag</li>
    }
  </ul>
</div>
</text>)(null);
}

<div>
  @showThing("Name", 1, new[] { "a", "b", "c", })
</div>
```

Can someone check for me if this still works in Core? Thanks.

Now, I do have a library of `@helper` which would be annoying to rewrite. Also, I rely on the `App_Code/MyHelpers.cshtml` way of magically importing my helpers to multiple razor files. If that really went away in the Core version, it’ll be a pain to switch to the new way of sharing code. But maybe one of the strongly typed code reuse options listed earlier would be sufficient, would require refactoring my stuff, and it would probably result in me cleaning up a lot of sloppy stuff. However, it would certainly be easier for to update me if `@helper` and `App_Code`-style sharing were still available…

**Comment by grokky1** [comment]
@binki How if at all did you manage to place those helpers in separate files?

**Comment by rynowak** [comment]
We have no plans to do this

**Comment by smdrager** [comment]
@Eilon 

@Helper had the benefit of being extremely fast, razor syntax, strongly typed, and as many helpers as you wanted in each file. The primary cons were having to put it in App_Code to share across views, having to create static classes for helpers like HtmlHelper and UrlHelper, and some oddities in formatting/indenting with Visual Studio.

The applications my company has created has hundreds of @Helper functions in App_Code. There is no way to have a high level of speed, razor syntax, and strongly typed calls without them. If they are removed we will have to move to a ton of compiled methods returning IHtmlString built from StringBuilders (or a custom type of chaining TagBuilders) which is truly a step down from @Helper.

I updated your comparison of other options to @Helper.

1. Use a view component
**Pros:** None (over @Helper)
**Cons:** An extra CSHTML file for each set of HTML

2. Use an HTML partial view
**Pros:** None (over @Helper)
**Cons:** No strongly-typed parameters, HORRIBLY slow (100x slower than @Helper) 

3. Use an HTML helper or Razor tag helper
**Pros:** None (over @Helper)
**Cons:** Everything is code, so no Razor syntax within the helper

**Comment by atifaziz** [comment]
@NTaylorMullen [said](https://github.com/aspnet/Razor/issues/715#issuecomment-198468516):

> This still does not solve the problem of a parameterized Razor code. What if there are two or three parameters of different types?

>>As unfortunate as it might be you'd need to provide some sort of poco object in that case. A potential, less clean solution would be using variables from outside of the `@<p>...</p>`:

The example given was:

```
var firstName = "John";
var lastName = "Doe";

Func<int, IHtmlContent> person = @<p>@firstName @lastName is @item years old.</p>;

@person(30)
```

But how about using a C# tuple to ship several values into the first argument? As in:

```
Func<(string FirstName, string LastName, int Age), IHtmlContent> person =
    @<p>@item.FirstName @item.LastName is @item.Age years old.</p>;

@person(("John", "Doe", 42))
```

The only odd looking bit may be the double parentheses needed at the call sites but that's far better than having to rely on closures for parameterization.

**Comment by atifaziz** [comment]
In addition to using C# tuples, [as shown above](https://github.com/aspnet/Razor/issues/715#issuecomment-328084286), using local C# functions can also make it simpler to pass multiple arguments:

```
IHtmlContent Render<T>(Func<T, IHtmlContent> helper, T item = default(T)) =>
    helper(item);

Func<object, IHtmlContent> Person(string fn, string ln, int age) =>
    @<p>@fn @ln is @age years old.</p>;

@Render(Person("John", "Doe", 42))
```

`Render` above is only for cosmetics as `@Person("John", "Doe", 42)(null)` would look ugly.

**Comment by duncansmart** [comment]
What are the extensibility points for the community to add `@helper` support?

**Comment by NTaylorMullen** [comment]
@duncansmart take a look at how we built some of MVCs directives:
- [Inject directive](https://github.com/aspnet/Razor/blob/0a76ad70172f29721173b07ef2707bc558a4c367/src/Microsoft.AspNetCore.Mvc.Razor.Extensions/InjectDirective.cs)
- [Model directive](https://github.com/aspnet/Razor/blob/0a76ad70172f29721173b07ef2707bc558a4c367/src/Microsoft.AspNetCore.Mvc.Razor.Extensions/ModelDirective.cs).

You'll run into some issues with getting the exact syntax of the old `@helper` directive but you'll be able to get somewhat close.

**Comment by jahanalem** [comment]
How can I compatible this code with **asp.net core 2** ?

`//Recursive function for rendering child nodes for the specified node

@helper CreateNavigation(int parentId, int depthNavigation, int currentPageId)
{
   @MyHelpers.Navigation(parentId, depthNavigation, currentPageId);
}

@helper Navigation(int parentId, int depthNavigation, int currentPageId)
{

   if ()
   {
       if ()
       {
         <ul style="">
            @foreach ()
            {
               if ()
                {
                   <li class="">
                      @Navigation(child.Id, depthNavigation, currentPageId)
                   </li>
                }
             }
         </ul>
      }
   }
}`

**Comment by NTaylorMullen** [comment]
@jahanalem's, @atifaziz  suggestions should be enough to massage your code back into a similar looking state.

**Comment by johnhargrove** [comment]
This issue should be re-opened as it is still not really addressed. I have been using Core for a couple years now and I run into this limitation all the time.

- ViewComponents are not a replacement. Writing HTML inside of a bunch of string concatenations buried inside some C# classes well away from the view is a step backward, and honestly it should be avoided when possible. 
- Partials aren't easily parameterized. Other limitations are already well-covered in this thread.
- The various suggested workarounds above are syntactic nightmares.

**Comment by frogcrush** [comment]
I agree. The limitation I have right now is you can't (sensibly) make recursive functions. Partials are simply too slow for this, especially if you're generating something like a menu or a tree view.

**Comment by ShadowDancer** [comment]
@NTaylorMullen 
Your solution is so ugly, that I believe that We should pretend it doesn't exist. Especially compared to so nice `@helper`.

**Comment by plus1319** [comment]
how about this code in mvc5 , without @helper i can't implement this perfectly
Recursive Category
@{
     AddOption(int? parentId)
        {
            foreach(var item in Model.Where(p=> p.ParentId == parentiId).ToList())
            {
                <option value="@item.Id">@item.Name</option>
                AddOption(item.Id);
            }
        }
} 
<select>
    <option value="">Main Category</option>
    @AddOption(null)
</select>

**Comment by sveins12** [comment]
It's a disaster.

**Comment by m0sa** [comment]
@rjgotten nice idea... I've added it to my benchmarks: 

https://github.com/m0sa/Mvc

Looks like closing over the helper arguments performs a lot better than actually passing them in as the model to the template!

**Comment by duncansmart** [comment]
That's a nice workaround improvement @rjgotten. Some issues with it after trying it out. The helper vars need to be declared before they're used (with `@helper` you can declare them at the bottom of the view for example). Also the intellisense for the args suffers as they're just genetic `Func<>`s

![image](https://user-images.githubusercontent.com/576343/49015964-871f1500-f17c-11e8-9bf6-98b20f389c12.png)

**Comment by rjgotten** [comment]
@duncansmart 
Hmm.... in that case, I guess you could still go the local function route and pull the helper resolution part _inside_ the definition of the local function.

E.g.

```csharp
public static HelperExtensions
{
  public static IHtmlContent Helper(this RazorPageBase, Func<object, IHtmlContent> body)
  {
    return body(null);
  }
}
```

And then consume as

```
IHtmlContent Person(string fn, string ln, int age) => 
  this.Helper(@<p>@fn @ln is @age years old.</p>)

Person("John", "Doe", 42)

```


That should solve both the argument naming problem and the definition hoisting problem.

**Comment by duncansmart** [comment]
Yep, not bad 👍 

![image](https://user-images.githubusercontent.com/576343/49076911-f9056600-f231-11e8-9b22-2b48981fb5a1.png)

Still want `@helper` back though :-)

**Comment by rjgotten** [comment]
> Still want `@helper` back though :-)

Don't we all? :)
